polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -143,173 +138,66 @@ pub fn worker_entrypoint(
 				let prepare_job_kind = pvf.prep_kind();
 				let executor_params = (*pvf.executor_params()).clone();
 
-				// Conditional variable to notify us when a thread is done.
-				let condvar = thread::get_condvar();
-
-				// Run the memory tracker in a regular, non-worker thread.
-				#[cfg(any(target_os = "linux", feature = "jemalloc-allocator"))]
-				let condvar_memory = Arc::clone(&condvar);
-				#[cfg(any(target_os = "linux", feature = "jemalloc-allocator"))]
-				let memory_tracker_thread = std::thread::spawn(|| memory_tracker_loop(condvar_memory));
-
-				let cpu_time_start = ProcessTime::now();
-
-				// Spawn a new thread that runs the CPU time monitor.
-				let (cpu_time_monitor_tx, cpu_time_monitor_rx) = channel::<()>();
-				let cpu_time_monitor_thread = thread::spawn_worker_thread(
-					"cpu time monitor thread",
-					move || {
-						cpu_time_monitor_loop(
-							cpu_time_start,
+				let mut rusage_before = unsafe { mem::zeroed() };
+				let mut rusage_after = unsafe { mem::zeroed() };
+				if unsafe { libc::getrusage(libc::RUSAGE_CHILDREN, &mut rusage_before) } == -1 {
+					send_response(
+						&mut stream,
+						Err(PrepareError::Panic(format!(
+							"error getting children resource usage for worker pid {}",
+							worker_pid
+						))),
+					)
+					.await?;
+				};
+				let mut pipe_fds = [0; 2];
+				if unsafe { libc::pipe(pipe_fds.as_mut_ptr()) } == -1 {
```

Comment:
Indeed. Please minimize usage of `unsafe` - each unsafe should only be used after quite some research whether it is possible to go without. I actually remembered that we have a strict no `unsafe` policy in Polkadot, but I do find unsafes in substrate. In any case, each usage has to be justified, proven to be correct and should have test cases where applicable. 

We are bypassing the typesystem, so we need to have extra guard rails. It would be super bad if a feature which was meant to improve security ends up being the security vulnerability. This has to be avoided at all cost. If we end up having any non super trivial unsafe code, I would even consider not doing the whole thing as not worth it/being counter productive.

Reply:
I haven't looked too closely whether it is feasible here, but let's at least aim for getting rid of those unsafes. If any remains, it should be thoroughly justified.

Reply:
Wasn't aware of the unsafe policy in Polkadot. At least in substrate, it is indeed allowed, but always requires a proof: https://github.com/paritytech/polkadot-sdk/blob/master/docs/STYLE_GUIDE.md#style

We do a lot of low-level stuff in the PVF workers, so it's impossible to *completely* avoid unsafe. For example, `pivot_root` requires a few system calls to set up (and is a syscall itself). In this PR, after removing `getrusage` and abstracting over pipes, we will just be left with `fork` and `setrlimit`. It should be rather trivial to prove that these are used safely.

---

Actually, there is the [rustix crate](https://docs.rs/rustix/latest/rustix/) which provides a safe wrapper around a lot of syscalls (though not `fork` for some reason, unless I missed it). I didn't want to introduce another dependency, when we are trying to remove dependencies from the workers. But, the compiler should optimize away any unused syscalls, so I think we can use `rustix`, starting with `setrlimit`.

Reply:
That style guide was originally in the substrate repo. Does it now apply to the whole monorepo, including Polkadot? Where were the Polkadot unsafe guidelines documented?

Reply:
Not sure, I think Rob just told me at some point. Anyhow, it should be a last resort - if you can justify its usage (and you seem to do), I will take the blame. ;-)

If safe abstractions are available, please definitely use them!

Reply:
@Jpserrat can you please explain why you chose `nix` over e.g. `rustix`? (I haven't looked into the differences.) It is really nice to get rid of all the `unsafe` blocks, so good job! Maybe we can use `nix` elsewhere also.

Reply:
@Jpserrat ping. :) 

> can you please explain why you chose nix over e.g. rustix?

Reply:
sorry, I didn't see that. I searched for safe implementations and `nix` was the first one and it looked like a well known and reliable lib, is the `rustix` a better one?

Reply:
I am not sure how they are different, so I would just go with whichever one we already use. In particular I've [seen `rustix` used](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/executor/wasmtime/Cargo.toml#L41) in executor code. 

Looks like both are present in `Cargo.lock`, but running `cargo tree` in `pvf/prepare-worker` or `pvf/execute-worker` shows that `rustix` is present as a dependency and not `nix`. In the interest of reducing dependencies in these two crates (see https://github.com/paritytech/polkadot-sdk/issues/649), I would opt for `rustix`. It might make the resulting binaries a bit smaller.

Reply:
Looks like `rustix` doesn't provide `getrusage`. If  you haven't already made the change, let's just leave `nix` for now and try to reduce the dependencies later in #649.

Reply:
yeah, I was going to mention that. Sorry for taking this long to respond to you, this week was hard. I'm gonna take a look at your other comments tomorrow morning, thanks a lot Marcin.